### PR TITLE
*: add index  best-practice

### PR DESCRIPTION
**Pull request by crazycs520**
### What is changed, added or deleted? 
* Add best-practice doc for add index.

### What is the related PR or file link(s)? 

[google doc](https://docs.google.com/document/d/14wAHAWmNXPTqEbifLc6_A70HRTpyHTTMiNoD5PUDVag/edit#)

### Which version does your change affect? 
v3.0, v2.1, dev,

### CODE REVIEW

File: v3.0/best-practices/add-index.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+## 测试目的
+
+对比 TiDB 3.0 版本和 2.0 版本 `add Index` 的速度以及调参对比和建议。
+
+## 测试版本、时间、地点
+
+TiDB 版本：
+
+    - Release 3.0: `6e2d6c7aa7eba3ac4f3a5201f1a36bf534fa6298`
```
Comment (lilin90) [inline v3.0/best-practices/add-index.md]:
```suggestion
- Release 3.0：`6e2d6c7aa7eba3ac4f3a5201f1a36bf534fa6298`
```

File: v3.0/best-practices/add-index.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    - v2.0.6
```
Comment (lilin90) [inline v3.0/best-practices/add-index.md]:
```suggestion
- v2.0.6
```

File: v3.0/best-practices/add-index.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+500000+0 records in
+500000+0 records out
+4096000000 bytes (4.1 GB) copied, 1.06741 s, 3.8 GB/s
+```
+
+### 测试数据
+
+一共对 2 张表进行测试，每张表 2000 万行数据：
+
+    - 表 1 是宽表：`t_wide`，200 列， c0 ~ c65 是 `int` 类型， c66 ~ c132 是 `varchar(200)` 类型， c133 ~ c199 是 `timestamp` 类型。
```
Comment (lilin90) [inline v3.0/best-practices/add-index.md]:
```suggestion
- 表 1 是宽表：`t_wide`，200 列，c0 ~ c65 是 `int` 类型，c66 ~ c132 是 `varchar(200)` 类型，c133 ~ c199 是 `timestamp` 类型。
```

File: v3.0/best-practices/add-index.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    - 表 2 是窄表：`t_slim`， 10 列，c0 ~ c2 是 `int` 类型， c3 ~ c5 是 `varchar(200)` 类型， c6 ~ c9 是 `timestamp` 类型。
```
Comment (lilin90) [inline v3.0/best-practices/add-index.md]:
```suggestion
- 表 2 是窄表：`t_slim`，10 列，c0 ~ c2 是 `int` 类型，c3 ~ c5 是 `varchar(200)` 类型，c6 ~ c9 是 `timestamp` 类型。
```

**Comment by crazycs520** [comment]
@bb7133 PTAL

File: v3.0/benchmark/add-index.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+`alter table t add unique index idx_(c0)`
+
+![add-index-best-practices-02](/media/add-index-best-practices-02.png)
+
+### 测试结论
+
+在相同的表结构下，测试 200W 数据数据的 add index 时间，基本和 2000W 数据的 add index 的时间成 10 倍的关系。
+
+以此类推， TiDB v3.0 版本 1亿 数据在类似的硬件，且无其他负载情况下， add index 消耗的时间大致为 344 * 5 = 1720 s = 29 min。有其他负载情况下 add index 操作会更慢一些，因为 add index 是低优先级操作。
```
Comment (bb7133) [inline v3.0/benchmark/add-index.md]:
```suggestion
以此类推，TiDB v3.0 版本 1 亿数据在类似的硬件，且无其他负载情况下，add index 消耗的时间大致为 344s * 5 = 1720s = 29 min。考虑到 add index 是低优先级操作，有其他负载情况下 add index 操作会更慢一些。
```

File: v3.0/best-practices/add-index.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+- Release 3.0：`6e2d6c7aa7eba3ac4f3a5201f1a36bf534fa6298`
+
+时间：2018 年 12 月 11 日
+
+地点：北京
+
+## 测试环境
+
+4 台AWS 机器，1 台机器部署 TiDB，PD, 监控，另外 3 台部署 TiKV，机器配置均为  E5-2686 v4 @ 2.30GHz，8核 59G，1.7T  本地 SSD。
```
Comment (bb7133) [inline v3.0/best-practices/add-index.md]:
```suggestion
4 台 AWS 机器，1 台机器部署 TiDB、PD 以及监控，另外 3 台部署 TiKV，机器配置均为  E5-2686 v4 @ 2.30GHz，8核 59G，1.7T  本地 SSD。
```

File: v3.0/best-practices/add-index.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+## 参数调整测试
+
+目前 TiDB 3.0 影响 add index 速度的参数有 2 个：
+
+* `tidb_ddl_reorg_worker_cnt` ：add index 的并发度，默认是 4
+* `tidb_ddl_reorg_batch_size` ：每次 add index 时的 batch 大小，默认是 256
+
+注：3.0.2 以及 2.0.16 之前，`tidb_ddl_reorg_worker_cnt` 和 `tidb_ddl_reorg_batch_size` 的默认值分别是 16 ，1024。
+
+以下测试是对各个参数单独设置 看其对 `add index` 的影响。
```
Comment (bb7133) [inline v3.0/best-practices/add-index.md]:
```suggestion
以下测试是对各个参数单独设置看其对 `add index` 的影响。
```

File: v3.0/best-practices/add-index.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+| 4608       | 1          | 138            | 90             |
+| 5120       | 1          | 139            | 87             |
+| 5632       | 1          | 138            | 85             |
+| 6144       | 1          | 137            | 87             |
+
+![add-index-best-practices-04](/media/add-index-best-practices-04.png)
+
+### 测试结论
+
+在无其他负载情况下，想让 add index 尽快完成，可以将 `tidb_ddl_reorg_worker_cnt` 和 `tidb_ddl_reorg_batch_size` 适当调大，比如 16，1024, 甚至 20，2048等。
```
Comment (bb7133) [inline v3.0/best-practices/add-index.md]:
```suggestion
在无其他负载情况下，想让 add index 尽快完成，可以将 `tidb_ddl_reorg_worker_cnt` 和 `tidb_ddl_reorg_batch_size` 适当调大，比如 16 和 1024。
```

File: v3.0/best-practices/add-index.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+在有其他负载情况下，想让 add index 尽量不影响其他业务，可以将 `tidb_ddl_reorg_worker_cnt` 和 `tidb_ddl_reorg_batch_size` 适当调小，比如 4, 256。
```
Comment (bb7133) [inline v3.0/best-practices/add-index.md]:
```suggestion
在有其他负载情况下，想让 add index 尽量不影响其他业务，可以将 `tidb_ddl_reorg_worker_cnt` 和 `tidb_ddl_reorg_batch_size` 适当调小，比如 4 和 256。
```

File: v3.0/best-practices/add-index.md [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+另外，在 `add index` 的同时，如果有对该表有比较频繁的 `Update` 操作，建议调小 `tidb_ddl_reorg_batch_size` ，比如 `128`，否则可能导致 `add index` 和 `update` 的事务冲突导致重试。
```
Comment (bb7133) [inline v3.0/best-practices/add-index.md]:
```suggestion
另外，在 `add index` 的同时，如果有对该表有比较频繁的 `update` 操作，建议调小 `tidb_ddl_reorg_batch_size` ，比如 `128`，否则可能导致 `add index` 和 `update` 的事务冲突导致重试。
```

File: v3.0/best-practices/add-index.md [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+目前 TiDB 3.0 影响 add index 速度的参数有 2 个：
+
+* `tidb_ddl_reorg_worker_cnt` ：add index 的并发度，默认是 4
+* `tidb_ddl_reorg_batch_size` ：每次 add index 时的 batch 大小，默认是 256
+
+注：3.0.2 以及 2.0.16 之前，`tidb_ddl_reorg_worker_cnt` 和 `tidb_ddl_reorg_batch_size` 的默认值分别是 16 ，1024。
+
+以下测试是对各个参数单独设置 看其对 `add index` 的影响。
+
+注：为加速测试，下面测试的数据量为 200W。
```
Comment (bb7133) [inline v3.0/best-practices/add-index.md]:
```suggestion
注：为加速测试，下面测试的数据量为 200 万。
```
请注意改正文章中类似的用法

**Comment by TomShawn** [comment]
@crazycs520 When the content is ready, please also update v3.1 as well, thanks~

**Comment by lilin90** [comment]
@crazycs520 Please address comments. Thanks!

**Comment by crazycs520** [comment]
@bb7133 PTAL

File: v3.0/benchmark/add-index.md [hunk 12]
```diff
@@ -0,0 +1,106 @@
+---
+title: Add Index Benchmark
+category: benchmark
+---
+
+# Add Index Benchmark
+
+## 测试目的
+
+对比 TiDB 3.0 版本和 2.0 版本 `add Index` 的速度。
```
Comment (bb7133) [inline v3.0/benchmark/add-index.md]:
```suggestion
对比 TiDB 3.0 版本和 2.0 版本 `Add Index` 的速度。
```

**Comment by crazycs520** [comment]
@lilin90 PTAL

File: v3.0/best-practices/add-index.md [hunk 13]
```diff
@@ -0,0 +1,122 @@
+---
```
Comment (lilin90) [inline v3.0/best-practices/add-index.md]: @anotherrachel PTAL~ What do you think of this "best practices"?

Comment (anotherrachel) [reply v3.0/best-practices/add-index.md]: This doc hardly counts as "best practices", neither form nor content. Suggest change the PR title. @lilin90

File: v3.0/benchmark/add-index.md [hunk 14]
```diff
@@ -0,0 +1,106 @@
+---
+title: Add Index Benchmark
```
Comment (lilin90) [inline v3.0/benchmark/add-index.md]:
```suggestion
title: Add Index 性能测试 - v3.0 对比 v2.0
```

File: v3.0/benchmark/add-index.md [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+category: benchmark
+---
+
+# Add Index Benchmark
```
Comment (lilin90) [inline v3.0/benchmark/add-index.md]:
```suggestion
# `Add Index` 性能测试 - v3.0 对比 v2.0
```

File: v3.0/best-practices/add-index.md (hunk 13 above)
Comment (lilin90) [reply v3.0/best-practices/add-index.md]: @crazycs520 这个不能算最佳实践哈，而且跟前面一个文档有重复内容，感觉也是个测试？😂或者直接把结论加到 add index 语句文档里？

File: v3.0/benchmark/add-index.md [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+- Release 3.0：`6e2d6c7aa7eba3ac4f3a5201f1a36bf534fa6298`
+- v2.0.6
+
+时间：2018 年 12 月 11 日
+
+地点：北京
+
+## 测试环境
+
+4 台AWS 机器，1 台机器部署 TiDB，PD, 监控，另外 3 台部署 TiKV，机器配置均为 E5-2686 v4 @ 2.30GHz，8核 59G，1.7T 本地 SSD。
```
Comment (lilin90) [inline v3.0/benchmark/add-index.md]:
```suggestion
4 台 AWS 机器，1 台机器部署 TiDB、PD、监控，另外 3 台部署 TiKV，机器配置均为 E5-2686 v4 @ 2.30GHz，8 核 59G，1.7T 本地 SSD。
```

**Comment by yikeke** [comment]
Please decide on the unresolved comment. This PR is too old. @crazycs520 @lilin90

**Comment by lilin90** [comment]
> Please decide on the unresolved comment. This PR is too old. @crazycs520 @lilin90

Exactly! @crazycs520 Could you please resolve this comment https://github.com/pingcap/docs-cn/pull/1886#discussion_r380651964?

**Comment by yikeke** [comment]
Closed because this PR has no progress for a long time. @crazycs520

**Pull request closed without merging**